src/tree/node.rs

```diff
@@ -0,0 +1,384 @@
+        Node {key: key, value: value, left: None, right: None, level: 1 }
+    }
+
+    // Remove left horizontal link by rotating right
+    fn skew(&mut self) {
+        if self.left.as_ref().map_or(false, |x| x.level == self.level) {
+            let mut save = self.left.take().unwrap();
+            swap(&mut self.left, &mut save.right); // save.right now None
+            swap(self, &mut save);
+            self.right = Some(save);
+        }
+    }
+
+    // Remove dual horizontal link by rotating left and increasing level of
+    // the parent
+    fn split(&mut self) {
+        if self.right.as_ref().map_or(false,
+          |x| x.right.as_ref().map_or(false, |y| y.level == self.level)) {
+            let mut save = self.right.take().unwrap();
+            swap(&mut self.right, &mut save.left); // save.left now None
+            save.level += 1;
+            swap(self, &mut save);
+            self.left = Some(save);
+        }
+    }
+}
+
+pub fn traverse<K, V, F>(mut node: &Option<Box<Node<K, V>>>, mut f: F)
+    -> &Option<Box<Node<K, V>>> where F: FnMut(&Node<K, V>) -> Ordering {
+
+    while let Some(ref r) = *node {
+        node = match f(r) {
+            Less => &r.left,
+            Greater => &r.right,
+            Equal => break,
+        };
+    }
+
+    node
+}
+
+pub fn traverse_mut<K, V, F>(node: &mut Option<Box<Node<K, V>>>, f: F)
+    -> &mut Option<Box<Node<K, V>>> where F: FnMut(&Node<K, V>) -> Ordering {
+
+    let node = traverse(node, f);
+    unsafe { mem::transmute(node) }
+}
+
+pub fn max<K, V>(node: &Node<K, V>) -> Ordering {
+    if node.right.is_some() { Greater } else { Equal }
+}
+
+pub fn min<K, V>(node: &Node<K, V>) -> Ordering {
+    if node.left.is_some() { Less } else { Equal }
+}
+
+pub fn insert<K, V, C>(node: &mut Option<Box<Node<K, V>>>, key: K, value: V, cmp: &C)
+    -> Option<V> where C: Compare<K> {
+
+    match *node {
+        Some(ref mut save) => {
+            let old_value = match cmp.compare(&key, &save.key) {
+                Less => insert(&mut save.left, key, value, cmp),
+                Greater => insert(&mut save.right, key, value, cmp),
+                Equal => {
+                    save.key = key;
+                    return Some(replace(&mut save.value, value));
+                }
+            };
+
+            save.skew();
+            save.split();
+            old_value
+        }
+        None => {
+            *node = Some(box Node::new(key, value));
+            None
+        }
+    }
+}
+
+pub fn remove<K, V, C, Q: ?Sized>(node: &mut Option<Box<Node<K, V>>>, key: &Q, cmp: &C)
+    -> Option<V> where C: Compare<Q, K> {
+
+    fn heir_swap<K, V>(node: &mut Box<Node<K, V>>,
+                            child: &mut Option<Box<Node<K, V>>>) {
+        // *could* be done without recursion, but it won't borrow check
+        for x in child.iter_mut() {
+            if x.right.is_some() {
+                heir_swap(node, &mut x.right);
```

Comment:
I'll investigate this. The recursive version was carried over from the previous code.